server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -171,13 +171,31 @@ private boolean shouldSync(boolean didRefresh, boolean skipPrimaryTermCheck) {
             // When the shouldSync is called the first time, then 1st condition on primary term is true. But after that
             // we update the primary term and the same condition would not evaluate to true again in syncSegments.
             // Below check ensures that if there is commit, then that gets picked up by both 1st and 2nd shouldSync call.
-            || isRefreshAfterCommitSafe();
+            || isRefreshAfterCommitSafe()
+            || isRemoteSegmentStoreInSync() == false;
         if (shouldSync || skipPrimaryTermCheck) {
             return shouldSync;
         }
         return this.primaryTerm != indexShard.getOperationPrimaryTerm();
     }
 
+    /**
+     * Checks if all files present in local store are uploaded to remote store or part of excluded files.
+     *
+     * Different from IndexShard#isRemoteSegmentStoreInSync as it uses files uploaded cache in RemoteDirectory
+     * And it doesn't make a remote store call.
+     *
+     * @return true iff all the local files are uploaded to remote store.
+     */
+    boolean isRemoteSegmentStoreInSync() {
```

Comment:
we can reuse the IndexShard#isRemoteSegmentStoreInSync here. the only difference i see is that we are making a remote call to fetch latest metadata. Apart from that, both methods are using the files uploaded cache in RemoteDirectory.

To skip the remote latest metadata part we can create a overloaded method maybe?

Reply:
In `IndexShard`, we cannot know if `isRemoteSegmentStoreInSync` without reading the latest metadata . So we cannot unify this in `IndexShard` .